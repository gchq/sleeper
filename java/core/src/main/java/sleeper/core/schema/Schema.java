/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.schema;

import sleeper.core.schema.type.PrimitiveType;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.TreeSet;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * A Schema describes the fields found in a particular table in a Sleeper instance.
 */
public class Schema {
    private final List<Field> rowKeyFields;
    private final List<Field> sortKeyFields;
    private final List<Field> valueFields;

    private Schema(Builder builder) {
        rowKeyFields = validateRowKeys(builder.rowKeyFields);
        sortKeyFields = validateSortKeys(builder.sortKeyFields);
        valueFields = validateValueKeys(builder.valueFields);
        validateNoDuplicates(streamAllFields(rowKeyFields, sortKeyFields, valueFields));
    }

    public static Builder builder() {
        return new Builder();
    }

    /**
     * Creates an instance of this class using a schema file.
     *
     * @param  schemaPath  the path to the schema file
     * @return             an instance of this class
     * @throws IOException if an error occurs reading from the file
     */
    public static Schema load(Path schemaPath) throws IOException {
        return loadFromString(Files.readString(schemaPath));
    }

    /**
     * Creates an instance of this class using a JSON string.
     *
     * @param  schemaJson the JSON string representing a schema
     * @return            an instance of this class
     */
    public static Schema loadFromString(String schemaJson) {
        return new SchemaSerDe().fromJson(schemaJson);
    }

    public List<Field> getRowKeyFields() {
        return rowKeyFields;
    }

    public List<Field> getSortKeyFields() {
        return sortKeyFields;
    }

    public List<Field> getValueFields() {
        return valueFields;
    }

    public List<PrimitiveType> getRowKeyTypes() {
        return getMappedFields(rowKeyFields, f -> (PrimitiveType) f.getType());
    }

    public List<PrimitiveType> getSortKeyTypes() {
        return getMappedFields(sortKeyFields, f -> (PrimitiveType) f.getType());
    }

    public List<String> getRowKeyFieldNames() {
        return getMappedFields(rowKeyFields, Field::getName);
    }

    public List<String> getSortKeyFieldNames() {
        return getMappedFields(sortKeyFields, Field::getName);
    }

    public List<String> getValueFieldNames() {
        return getMappedFields(valueFields, Field::getName);
    }

    public List<String> getAllFieldNames() {
        return getMappedFields(streamAllFields(), Field::getName);
    }

    private <T> List<T> getMappedFields(List<Field> fields, Function<Field, T> mapping) {
        return getMappedFields(fields.stream(), mapping);
    }

    private <T> List<T> getMappedFields(Stream<Field> fields, Function<Field, T> mapping) {
        return fields
                .map(mapping)
                .collect(Collectors.toUnmodifiableList());
    }

    public List<Field> getAllFields() {
        return streamAllFields().collect(Collectors.toUnmodifiableList());
    }

    /**
     * Streams all fields in the schema. This includes row key, sort key, and value fields.
     *
     * @return a stream of all fields
     */
    public Stream<Field> streamAllFields() {
        return streamAllFields(rowKeyFields, sortKeyFields, valueFields);
    }

    /**
     * Gets a field by name.
     *
     * @param  fieldName the name of the field
     * @return           a field, or an empty optional if the field does not exist
     */
    public Optional<Field> getField(String fieldName) {
        return streamAllFields()
                .filter(f -> f.getName().equals(fieldName))
                .findFirst();
    }

    /**
     * Saves this schema to a file.
     *
     * @param  path        the path of the file
     * @throws IOException if an error occured when writing the file
     */
    public void save(Path path) throws IOException {
        Files.writeString(path, new SchemaSerDe().toJson(this));
    }

    @Override
    public String toString() {
        return "Schema{" + "rowKeyFields=" + rowKeyFields + ", sortKeyFields=" + sortKeyFields + ", valueFields=" + valueFields + '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Schema schema = (Schema) o;

        return Objects.equals(rowKeyFields, schema.rowKeyFields) &&
                Objects.equals(sortKeyFields, schema.sortKeyFields) &&
                Objects.equals(valueFields, schema.valueFields);
    }

    @Override
    public int hashCode() {
        return Objects.hash(rowKeyFields, sortKeyFields, valueFields);
    }

    /**
     * Builder to create a schema object.
     */
    public static final class Builder {
        private List<Field> rowKeyFields;
        private List<Field> sortKeyFields;
        private List<Field> valueFields;

        private Builder() {
        }

        /**
         * Sets the row key fields.
         *
         * @param  rowKeyFields the row key fields
         * @return              the builder
         */
        public Builder rowKeyFields(List<Field> rowKeyFields) {
            this.rowKeyFields = rowKeyFields;
            return this;
        }

        /**
         * Sets the row key fields.
         *
         * @param  rowKeyFields the row key fields
         * @return              the builder
         */
        public Builder rowKeyFields(Field... rowKeyFields) {
            return rowKeyFields(Arrays.asList(rowKeyFields));
        }

        /**
         * Sets the sort key fields.
         *
         * @param  sortKeyFields the sort key fields
         * @return               the builder
         */
        public Builder sortKeyFields(List<Field> sortKeyFields) {
            this.sortKeyFields = sortKeyFields;
            return this;
        }

        /**
         * Sets the sort key fields.
         *
         * @param  sortKeyFields the sort key fields
         * @return               the builder
         */
        public Builder sortKeyFields(Field... sortKeyFields) {
            return sortKeyFields(Arrays.asList(sortKeyFields));
        }

        /**
         * Sets the value fields.
         *
         * @param  valueFields the value fields
         * @return             the builder
         */
        public Builder valueFields(List<Field> valueFields) {
            this.valueFields = valueFields;
            return this;
        }

        /**
         * Sets the value fields.
         *
         * @param  valueFields the value fields
         * @return             the builder
         */
        public Builder valueFields(Field... valueFields) {
            return valueFields(Arrays.asList(valueFields));
        }

        public Schema build() {
            return new Schema(this);
        }
    }

    private static List<Field> validateRowKeys(List<Field> fields) {
        if (fields == null || fields.isEmpty()) {
            throw new IllegalArgumentException("Must have at least one row key field");
        }
        if (fields.stream().anyMatch(field -> !(field.getType() instanceof PrimitiveType))) {
            throw new IllegalArgumentException("Row key fields must have a primitive type");
        }
        return makeImmutable(fields);
    }

    private static List<Field> validateSortKeys(List<Field> fields) {
        if (fields == null) {
            return Collections.emptyList();
        }
        if (fields.stream().anyMatch(field -> !(field.getType() instanceof PrimitiveType))) {
            throw new IllegalArgumentException("Sort key fields must have a primitive type");
        }
        return makeImmutable(fields);
    }

    private static List<Field> validateValueKeys(List<Field> fields) {
        if (fields == null) {
            return Collections.emptyList();
        }
        return makeImmutable(fields);
    }

    private static List<Field> makeImmutable(List<Field> fields) {
        return Collections.unmodifiableList(new ArrayList<>(fields));
    }

    private static Stream<Field> streamAllFields(
            List<Field> rowKeyFields, List<Field> sortKeyFields, List<Field> valueFields) {
        return Stream.of(rowKeyFields, sortKeyFields, valueFields)
                .flatMap(List::stream);
    }

    private static void validateNoDuplicates(Stream<Field> fields) {
        Set<String> foundNames = new HashSet<>();
        Set<String> duplicates = new TreeSet<>();
        fields.forEach(field -> {
            boolean isNew = foundNames.add(field.getName());
            if (!isNew) {
                duplicates.add(field.getName());
            }
        });
        if (!duplicates.isEmpty()) {
            throw new IllegalArgumentException("Found duplicate field names: " + duplicates);
        }
    }
}
