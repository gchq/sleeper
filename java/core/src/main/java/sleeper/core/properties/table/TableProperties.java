/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.properties.table;

import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;

import sleeper.core.properties.PropertyGroup;
import sleeper.core.properties.SleeperProperties;
import sleeper.core.properties.SleeperPropertiesPrettyPrinter;
import sleeper.core.properties.SleeperPropertiesValidationReporter;
import sleeper.core.properties.SleeperPropertyIndex;
import sleeper.core.properties.instance.InstanceProperties;
import sleeper.core.schema.Schema;
import sleeper.core.schema.SchemaSerDe;
import sleeper.core.table.TableStatus;

import java.io.PrintWriter;
import java.util.Optional;
import java.util.Properties;
import java.util.stream.Collectors;

import static sleeper.core.properties.PropertiesUtils.loadProperties;
import static sleeper.core.properties.table.TableProperty.SCHEMA;
import static sleeper.core.properties.table.TableProperty.TABLE_ID;
import static sleeper.core.properties.table.TableProperty.TABLE_NAME;
import static sleeper.core.properties.table.TableProperty.TABLE_ONLINE;

/**
 * Contains values of the properties to configure a Sleeper table.
 */
public class TableProperties extends SleeperProperties<TableProperty> {

    // Schema is cached for faster access
    private Schema schema;
    private final InstanceProperties instanceProperties; // Used for default properties

    public TableProperties(InstanceProperties instanceProperties) {
        this.instanceProperties = instanceProperties;
    }

    public TableProperties(InstanceProperties instanceProperties, Properties properties) {
        super(properties);
        this.instanceProperties = instanceProperties;
        schema = loadSchema(properties);
    }

    /**
     * Creates a copy of the given table properties.
     *
     * @param  tableProperties the table properties
     * @return                 the copy
     */
    public static TableProperties copyOf(TableProperties tableProperties) {
        InstanceProperties instanceProperties = InstanceProperties.copyOf(tableProperties.instanceProperties);
        return new TableProperties(instanceProperties, loadProperties(tableProperties.saveAsString()));
    }

    /**
     * Creates and validates an instance of this class with the given property values.
     *
     * @param  instanceProperties the instance properties
     * @param  properties         the property values
     * @return                    the table properties
     */
    public static TableProperties createAndValidate(InstanceProperties instanceProperties, Properties properties) {
        TableProperties tableProperties = new TableProperties(instanceProperties, properties);
        tableProperties.validate();
        return tableProperties;
    }

    /**
     * Creates an instance with different property values than the given instance. Performs no validation, and uses the
     * same instance properties as the given instance.
     *
     * @param  tableProperties the original table properties
     * @param  newProperties   the property values
     * @return                 the new table properties
     */
    public static TableProperties recreateWithoutValidation(TableProperties tableProperties, Properties newProperties) {
        return new TableProperties(tableProperties.instanceProperties, newProperties);
    }

    private static Schema loadSchema(Properties properties) {
        return Optional.ofNullable(properties.getProperty(SCHEMA.getPropertyName()))
                .map(Schema::loadFromString)
                .orElse(null);
    }

    @Override
    protected void init() {
        String schemaProperty = get(TableProperty.SCHEMA);
        if (schemaProperty != null) {
            schema = Schema.loadFromString(schemaProperty);
        }
        super.init();
    }

    @Override
    public void validate(SleeperPropertiesValidationReporter reporter) {
        super.validate(reporter);
    }

    @Override
    public String get(TableProperty property) {
        return compute(property, value -> property.computeValue(value, instanceProperties, this));
    }

    public Schema getSchema() {
        return schema;
    }

    /**
     * Sets the schema of the Sleeper table.
     *
     * @param schema the schema
     */
    public void setSchema(Schema schema) {
        this.schema = schema;
        set(TableProperty.SCHEMA, new SchemaSerDe().toJson(schema));
    }

    @Override
    public SleeperPropertyIndex<TableProperty> getPropertiesIndex() {
        return TableProperty.Index.INSTANCE;
    }

    @Override
    protected SleeperPropertiesPrettyPrinter<TableProperty> getPrettyPrinter(PrintWriter writer) {
        return createPrettyPrinter(writer);
    }

    /**
     * Creates a printer to be used to display all table properties.
     *
     * @param  writer the writer to write to
     * @return        the pretty printer
     */
    public static SleeperPropertiesPrettyPrinter<TableProperty> createPrettyPrinter(PrintWriter writer) {
        return SleeperPropertiesPrettyPrinter.builder()
                .properties(TableProperty.getAll(), TablePropertyGroup.getAll())
                .writer(writer).build();
    }

    /**
     * Creates a printer to be used to display table properties in a given group.
     *
     * @param  writer the writer to write to
     * @param  group  the group to display
     * @return        the pretty printer
     */
    public static SleeperPropertiesPrettyPrinter<TableProperty> createPrettyPrinterWithGroup(
            PrintWriter writer, PropertyGroup group) {
        return SleeperPropertiesPrettyPrinter.builder()
                .sortedProperties(TableProperty.getAll().stream()
                        .filter(property -> property.getPropertyGroup().equals(group))
                        .collect(Collectors.toList()))
                .writer(writer).hideUnknownProperties(true).build();
    }

    public TableStatus getStatus() {
        return TableStatus.uniqueIdAndName(get(TABLE_ID), get(TABLE_NAME), getBoolean(TABLE_ONLINE));
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        TableProperties that = (TableProperties) o;

        return new EqualsBuilder()
                .appendSuper(super.equals(o))
                .append(instanceProperties, that.instanceProperties)
                .isEquals();
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder(17, 37)
                .appendSuper(super.hashCode())
                .append(instanceProperties)
                .toHashCode();
    }
}
