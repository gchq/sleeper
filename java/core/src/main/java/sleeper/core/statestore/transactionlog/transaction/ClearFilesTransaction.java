/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.statestore.transactionlog.transaction;

import sleeper.core.statestore.StateStoreException;
import sleeper.core.statestore.transactionlog.state.StateStoreFiles;

import java.time.Instant;
import java.util.Objects;

/**
 * A transaction to delete all files in the state store.
 */
public class ClearFilesTransaction implements FileReferenceTransaction {

    @Override
    public void validate(StateStoreFiles stateStoreFiles) throws StateStoreException {
    }

    @Override
    public void apply(StateStoreFiles stateStoreFiles, Instant updateTime) {
        stateStoreFiles.clear();
    }

    @Override
    public int hashCode() {
        return Objects.hash();
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof ClearFilesTransaction;
    }

    @Override
    public String toString() {
        return "ClearFilesTransaction{}";
    }
}
