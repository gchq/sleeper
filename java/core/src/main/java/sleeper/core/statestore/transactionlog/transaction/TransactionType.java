/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.statestore.transactionlog.transaction;

/**
 * A marker for the type of a transaction. Used when storing a log entry for a transaction. Stabilises mapping between
 * a serialised transaction type and the class of the transaction, in case the class is renamed or moved.
 */
public enum TransactionType {

    ADD_FILES(AddFilesTransaction.class),
    ASSIGN_JOB_IDS(AssignJobIdsTransaction.class),
    CLEAR_FILES(ClearFilesTransaction.class),
    DELETE_FILES(DeleteFilesTransaction.class),
    INITIALISE_PARTITIONS(InitialisePartitionsTransaction.class),
    REPLACE_FILE_REFERENCES(ReplaceFileReferencesTransaction.class),
    SPLIT_FILE_REFERENCES(SplitFileReferencesTransaction.class),
    SPLIT_PARTITION(SplitPartitionTransaction.class);

    private final Class<? extends StateStoreTransaction<?>> type;

    TransactionType(Class<? extends StateStoreTransaction<?>> type) {
        this.type = type;
    }

    public Class<? extends StateStoreTransaction<?>> getType() {
        return type;
    }

    /**
     * Retrieves the type of a transaction.
     *
     * @param  transaction the transaction
     * @return             the type
     */
    public static TransactionType getType(StateStoreTransaction<?> transaction) {
        for (TransactionType type : values()) {
            if (type.getType().isInstance(transaction)) {
                return type;
            }
        }
        throw new IllegalArgumentException("Transaction type not recognised: " + transaction.getClass());
    }
}
