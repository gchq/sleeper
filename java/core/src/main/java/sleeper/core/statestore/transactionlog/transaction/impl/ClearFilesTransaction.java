/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.statestore.transactionlog.transaction.impl;

import sleeper.core.statestore.StateStore;
import sleeper.core.statestore.StateStoreException;
import sleeper.core.statestore.transactionlog.AddTransactionRequest;
import sleeper.core.statestore.transactionlog.state.StateStoreFiles;
import sleeper.core.statestore.transactionlog.transaction.FileReferenceTransaction;

import java.time.Instant;
import java.util.Objects;

/**
 * Clears all file data from the file reference store. Note that this does not delete any of the actual files.
 */
public class ClearFilesTransaction implements FileReferenceTransaction {

    @Override
    public void validate(StateStoreFiles stateStoreFiles) throws StateStoreException {
    }

    /**
     * Commit this transaction directly to the state store without going to the commit queue. This will throw any
     * validation exceptions immediately, even if they wouldn't be as part of an asynchronous commit.
     *
     * @param  stateStore          the state store
     * @throws StateStoreException if the update fails
     */
    public void synchronousCommit(StateStore stateStore) throws StateStoreException {
        stateStore.addFilesTransaction(AddTransactionRequest.withTransaction(this).build());
    }

    @Override
    public void apply(StateStoreFiles stateStoreFiles, Instant updateTime) {
        stateStoreFiles.clear();
    }

    @Override
    public int hashCode() {
        return Objects.hash();
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof ClearFilesTransaction;
    }

    @Override
    public String toString() {
        return "ClearFilesTransaction{}";
    }
}
