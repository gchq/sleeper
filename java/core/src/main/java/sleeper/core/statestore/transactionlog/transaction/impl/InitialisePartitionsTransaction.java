/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.statestore.transactionlog.transaction.impl;

import sleeper.core.partition.Partition;
import sleeper.core.statestore.transactionlog.state.StateStorePartitions;
import sleeper.core.statestore.transactionlog.transaction.PartitionTransaction;

import java.time.Instant;
import java.util.List;
import java.util.Objects;

/**
 * A transaction to set all partitions in a Sleeper table. This should specify the whole partition tree. Any partitions
 * that were present before will be deleted.
 */
public class InitialisePartitionsTransaction implements PartitionTransaction {

    private final List<Partition> partitions;

    public InitialisePartitionsTransaction(List<Partition> partitions) {
        this.partitions = partitions;
    }

    @Override
    public void validate(StateStorePartitions stateStorePartitions) {
    }

    @Override
    public void apply(StateStorePartitions stateStorePartitions, Instant updateTime) {
        stateStorePartitions.clear();
        partitions.forEach(stateStorePartitions::put);
    }

    @Override
    public int hashCode() {
        return Objects.hash(partitions);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (!(obj instanceof InitialisePartitionsTransaction)) {
            return false;
        }
        InitialisePartitionsTransaction other = (InitialisePartitionsTransaction) obj;
        return Objects.equals(partitions, other.partitions);
    }

    @Override
    public String toString() {
        return "InitialisePartitionsTransaction{partitions=" + partitions + "}";
    }
}
