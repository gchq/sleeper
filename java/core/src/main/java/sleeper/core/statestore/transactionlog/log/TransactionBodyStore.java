/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.statestore.transactionlog.log;

import sleeper.core.properties.table.TableProperties;
import sleeper.core.statestore.commit.StateStoreCommitRequest;
import sleeper.core.statestore.transactionlog.AddTransactionRequest;
import sleeper.core.statestore.transactionlog.TransactionLogStateStore;
import sleeper.core.statestore.transactionlog.transaction.StateStoreTransaction;
import sleeper.core.statestore.transactionlog.transaction.TransactionType;

import java.time.Instant;
import java.util.UUID;

import static sleeper.core.properties.table.TableProperty.TABLE_ID;

/**
 * A store of the bodies of transactions that will be referenced in a transaction log. Used by
 * {@link TransactionLogStateStore} for transactions that are too large to be held directly in
 * {@link TransactionLogStore}.
 */
public interface TransactionBodyStore {

    /**
     * Stores a transaction at a given location.
     *
     * @param key         the object key in the data bucket
     * @param tableId     the ID of the Sleeper table the transaction is for
     * @param transaction the transaction
     */
    void store(String key, String tableId, StateStoreTransaction<?> transaction);

    /**
     * Stores a transaction if it is too large to fit in the associated log store.
     *
     * @param  tableId the ID of the Sleeper table the transaction is for
     * @param  request the request to add the transaction
     * @return         the new request, identical to the original but with a pointer to the new transaction entry if it
     *                 was uploaded
     */
    default AddTransactionRequest storeIfTooBig(String tableId, AddTransactionRequest request) {
        return request;
    }

    /**
     * Retrives a transaction from a given location.
     *
     * @param  key             the object key in the data bucket
     * @param  tableId         the ID of the Sleeper table the transaction is for
     * @param  transactionType the type of the transaction
     * @return                 the transaction
     */
    <T extends StateStoreTransaction<?>> T getBody(String key, String tableId, TransactionType transactionType);

    /**
     * Retrives a transaction from commit request.
     *
     * @param  request the commit request
     * @return         the transaction
     */
    default <T extends StateStoreTransaction<?>> T getTransaction(StateStoreCommitRequest request) {
        return request.<T>getTransactionIfHeld()
                .orElseGet(() -> getBody(request.getBodyKey(), request.getTableId(), request.getTransactionType()));
    }

    /**
     * Creates an object key for a new transaction file with a randomly generated filename. The file will not yet exist.
     *
     * @param  tableProperties the Sleeper table properties
     * @return                 the object key
     */
    static String createObjectKey(TableProperties tableProperties) {
        return createObjectKey(tableProperties.get(TABLE_ID));
    }

    /**
     * Creates an object key for a new transaction file with a randomly generated filename. The file will not yet exist.
     *
     * @param  tableId the Sleeper table ID
     * @return         the object key
     */
    static String createObjectKey(String tableId) {
        // Use a random UUID to avoid conflicting when another process is adding a transaction at the same time
        return createObjectKey(tableId, Instant.now(), UUID.randomUUID().toString());
    }

    /**
     * Creates an object key for a new transaction file with a randomly generated filename. The file will not yet exist.
     *
     * @param  tableId the Sleeper table ID
     * @param  now     the time now
     * @param  uuid    a random UUID
     * @return         the object key
     */
    static String createObjectKey(String tableId, Instant now, String uuid) {
        return tableId + "/statestore/transactions/" + now + "-" + uuid + ".json";
    }
}
