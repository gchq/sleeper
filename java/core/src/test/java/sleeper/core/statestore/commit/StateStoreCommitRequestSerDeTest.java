/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.statestore.commit;

import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.junit.jupiter.api.Test;

import sleeper.core.partition.PartitionsBuilder;
import sleeper.core.properties.instance.InstanceProperties;
import sleeper.core.properties.table.TableProperties;
import sleeper.core.properties.testutils.FixedTablePropertiesProvider;
import sleeper.core.schema.Schema;
import sleeper.core.schema.type.LongType;
import sleeper.core.statestore.transactionlog.InMemoryTransactionBodyStore;
import sleeper.core.statestore.transactionlog.TransactionBodyStore;
import sleeper.core.statestore.transactionlog.transaction.PartitionTransaction;
import sleeper.core.statestore.transactionlog.transaction.TransactionType;
import sleeper.core.statestore.transactionlog.transaction.impl.ClearFilesTransaction;
import sleeper.core.statestore.transactionlog.transaction.impl.InitialisePartitionsTransaction;

import java.time.Instant;

import static org.assertj.core.api.Assertions.assertThat;
import static sleeper.core.properties.table.TableProperty.TABLE_ID;
import static sleeper.core.properties.table.TableProperty.TABLE_NAME;
import static sleeper.core.properties.testutils.InstancePropertiesTestHelper.createTestInstanceProperties;
import static sleeper.core.properties.testutils.TablePropertiesTestHelper.createTestTableProperties;
import static sleeper.core.schema.SchemaTestHelper.schemaWithKey;

public class StateStoreCommitRequestSerDeTest {

    Schema schema = schemaWithKey("key", new LongType());
    TableProperties tableProperties = createTableProperties(schema);
    StateStoreCommitRequestSerDe serDe = new StateStoreCommitRequestSerDe(new FixedTablePropertiesProvider(tableProperties));
    TransactionBodyStore bodyStore = new InMemoryTransactionBodyStore();

    @Test
    void shouldSerialiseCompactionCommitInS3() {
        // Given
        String key = TransactionBodyStore.createObjectKey("test-table", Instant.parse("2025-01-14T15:30:00Z"), "test-transaction");
        StateStoreCommitRequest commitRequest = StateStoreCommitRequest.create(
                "test-table", key, TransactionType.REPLACE_FILE_REFERENCES);

        // When
        String json = serDe.toJsonPrettyPrint(commitRequest);

        // Then
        assertThat(serDe.fromJson(json)).isEqualTo(commitRequest);
        Approvals.verify(json, new Options().forFile().withExtension(".json"));
    }

    @Test
    void shouldSerialiseInitialisePartitionsInS3() {
        // Given
        String key = TransactionBodyStore.createObjectKey("test-table", Instant.parse("2025-01-14T15:30:00Z"), "test-transaction");
        StateStoreCommitRequest commitRequest = StateStoreCommitRequest.create(
                "test-table", key, TransactionType.INITIALISE_PARTITIONS);

        // When
        String json = serDe.toJsonPrettyPrint(commitRequest);

        // Then
        assertThat(serDe.fromJson(json)).isEqualTo(commitRequest);
        Approvals.verify(json, new Options().forFile().withExtension(".json"));
    }

    @Test
    void shouldSerialiseInitialisePartitionsDirectly() {
        // Given
        PartitionTransaction transaction = new InitialisePartitionsTransaction(new PartitionsBuilder(schema)
                .rootFirst("root")
                .splitToNewChildren("root", "L", "R", 50L)
                .buildList());
        StateStoreCommitRequest commitRequest = StateStoreCommitRequest.create("test-table", transaction);

        // When
        String json = serDe.toJsonPrettyPrint(commitRequest);

        // Then
        assertThat(serDe.fromJson(json)).isEqualTo(commitRequest);
        Approvals.verify(json, new Options().forFile().withExtension(".json"));
    }

    @Test
    void shouldSeraliseFileTransactionWithoutTableProperties() {
        // Given
        StateStoreCommitRequestSerDe serDe = StateStoreCommitRequestSerDe.forFileTransactions();
        StateStoreCommitRequest commitRequest = StateStoreCommitRequest.create("test-table", new ClearFilesTransaction());

        // When
        String json = serDe.toJson(commitRequest);

        // Then
        assertThat(serDe.fromJson(json)).isEqualTo(commitRequest);
    }

    private TableProperties createTableProperties(Schema schema) {
        InstanceProperties instanceProperties = createTestInstanceProperties();
        TableProperties tableProperties = createTestTableProperties(instanceProperties, schema);
        tableProperties.set(TABLE_ID, "test-table");
        tableProperties.set(TABLE_NAME, "test.table");
        return tableProperties;
    }

}
