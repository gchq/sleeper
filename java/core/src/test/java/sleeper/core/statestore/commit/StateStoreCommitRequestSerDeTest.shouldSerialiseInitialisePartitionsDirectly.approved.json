{
  "tableId": "test-table",
  "transactionType": "INITIALISE_PARTITIONS",
  "transaction": {
    "partitions": [
      {
        "partitionId": "root",
        "isLeafPartition": false,
        "childPartitionIds": [
          "L",
          "R"
        ],
        "region": {
          "key": {
            "min": -9223372036854775808,
            "minInclusive": true,
            "maxInclusive": false
          },
          "stringsBase64Encoded": true
        },
        "dimension": 0
      },
      {
        "partitionId": "L",
        "isLeafPartition": true,
        "parentPartitionId": "root",
        "childPartitionIds": [],
        "region": {
          "key": {
            "min": -9223372036854775808,
            "minInclusive": true,
            "max": 50,
            "maxInclusive": false
          },
          "stringsBase64Encoded": true
        },
        "dimension": -1
      },
      {
        "partitionId": "R",
        "isLeafPartition": true,
        "parentPartitionId": "root",
        "childPartitionIds": [],
        "region": {
          "key": {
            "min": 50,
            "minInclusive": true,
            "maxInclusive": false
          },
          "stringsBase64Encoded": true
        },
        "dimension": -1
      }
    ]
  }
}